services/audit/record.go

```diff
@@ -0,0 +1,512 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package audit
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	asymkey_model "code.gitea.io/gitea/models/asymkey"
+	audit_model "code.gitea.io/gitea/models/audit"
+	auth_model "code.gitea.io/gitea/models/auth"
+	git_model "code.gitea.io/gitea/models/git"
+	organization_model "code.gitea.io/gitea/models/organization"
+	perm_model "code.gitea.io/gitea/models/perm"
+	repository_model "code.gitea.io/gitea/models/repo"
+	secret_model "code.gitea.io/gitea/models/secret"
+	user_model "code.gitea.io/gitea/models/user"
+	webhook_model "code.gitea.io/gitea/models/webhook"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+type Event struct {
+	Action    audit_model.Action `json:"action"`
+	Actor     TypeDescriptor     `json:"actor"`
+	Scope     TypeDescriptor     `json:"scope"`
+	Target    TypeDescriptor     `json:"target"`
+	Message   string             `json:"message"`
+	Time      time.Time          `json:"time"`
+	IPAddress string             `json:"ip_address"`
+}
+
+func buildEvent(ctx context.Context, action audit_model.Action, actor *user_model.User, scope, target any, message string, v ...any) *Event {
+	return &Event{
+		Action:    action,
+		Actor:     typeToDescription(actor),
+		Scope:     scopeToDescription(scope),
+		Target:    typeToDescription(target),
+		Message:   fmt.Sprintf(message, v...),
+		Time:      time.Now(),
+		IPAddress: tryGetIPAddress(ctx),
+	}
+}
+
+func record(ctx context.Context, action audit_model.Action, actor *user_model.User, scope, target any, message string, v ...any) {
+	if !setting.Audit.Enabled {
+		return
+	}
+
+	e := buildEvent(ctx, action, actor, scope, target, message, v...)
+
+	if err := writeToFile(e); err != nil {
+		log.Error("Error writing audit event to file: %v", err)
+	}
+	if err := writeToDatabase(ctx, e); err != nil {
+		log.Error("Error writing audit event to database: %v", err)
```

Comment:
1.  > 2024/03/01 00:07:30 ...ices/audit/record.go:58:record() [E] Error writing audit event to database: NOT NULL constraint failed: audit_event.timestamp

Happened on startup, so probably the `gitea started` message
EDIT: Yep, I guessed correctly
> 2024/03/01 00:14:15 ...ices/audit/record.go:58:record() [E] Error writing audit event &{Action:system:startup Actor:{Type:user ID:-3 Object:0xc001828780} Scope:{Type:system ID:0 Object:<nil>} Target:{Type:system ID:0 Object:<nil>} Message:System started [Gitea 1.22.0+dev-913-g3b7298bf6a] Time:2024-03-01 00:14:15.169360476 +0100 CET m=+0.766367175 IPAddress:} to database: NOT NULL constraint failed: audit_event.timestamp

2.
```suggestion
		log.Error("Error writing audit event %+v to database: %v", e, err)
```

3. > 2024/03/01 00:15:53 cmd/web.go:197:serveInstalled() [W] Table audit_event has column timestamp but struct has not related field

is probably related/the reason